File: peer-fixtures/opentui/sb3_datatable/bin/sb3_datatable_benchmark.js
#!/usr/bin/env bun

import { execFileSync } from 'node:child_process'
import { readFileSync, mkdirSync, writeFileSync } from 'node:fs'
import { dirname } from 'node:path'
import { fileURLToPath } from 'node:url'

import {
  Sb3OpenTuiTableHarness,
  expectedCopiedRow,
  rowId,
  unsafeCopyTextCount,
  unsafeVisibleTextCount,
  visibleDataCapacity,
} from '../lib/table_app.js'

const SCHEMA_VERSION = 1
const PEER_ID = 'opentui'
const PEER_NAME = 'OpenTUI'
const PEER_VERSION = '0.3.1'
const PEER_URL = 'https://www.npmjs.com/package/@opentui/core'
const SCENARIO_ID = 'SB.3'
const DEFAULT_WARMUPS = 2
const DEFAULT_ITERATIONS = 5
const DEFAULT_ROWS = 100_000
const DEFAULT_COLUMNS = 120
const DEFAULT_TERMINAL_ROWS = 32

const options = parseArgs(process.argv.slice(2))

for (let index = 0; index < options.warmupIterations; index += 1) {
  await runSample(options)
}

const rssBefore = process.memoryUsage().rss
const samples = []
for (let index = 0; index < options.measuredIterations; index += 1) {
  samples.push(await runSample(options))
}
const runRssDeltaBytes = Math.max(0, process.memoryUsage().rss - rssBefore)
const artifact = buildArtifact(options, samples, runRssDeltaBytes)
const jsonText = `${JSON.stringify(artifact, null, 2)}\n`

if (options.outputPath) {
  mkdirSync(dirname(options.outputPath), { recursive: true })
  writeFileSync(options.outputPath, jsonText)
}

if (options.printJson) {
  process.stdout.write(jsonText)
} else {
  console.log('OpenTUI SB.3 DataTable fixture')
  console.log(`Rows: ${options.rows}`)
  console.log(`Iterations: ${options.measuredIterations}`)
  console.log(`pageMoveUs p95: ${artifact.metrics.pageMoveUs.p95}`)
  console.log(`copySelectedRowUs p95: ${artifact.metrics.copySelectedRowUs.p95}`)
  if (options.outputPath) console.log(`Saved ${options.outputPath}`)
}

async function runSample(options) {
  let app
  const rssBefore = process.memoryUsage().rss
  const mountStart = process.hrtime.bigint()
  app = await Sb3OpenTuiTableHarness.create({
    rowCount: options.rows,
    terminalColumns: options.terminalColumns,
    terminalRows: options.terminalRows,
  })
  const mountUs = elapsedUs(mountStart)
  const rssAfterMount = process.memoryUsage().rss

  try {
    const firstRenderStart = process.hrtime.bigint()
    const firstFrame = await app.render()
    const firstRenderUs = elapsedUs(firstRenderStart)

    const arrowStart = process.hrtime.bigint()
    app.arrowDown()
    await app.render()
    const arrowMoveUs = elapsedUs(arrowStart)

    const pageStart = process.hrtime.bigint()
    app.pageDown()
    await app.render()
    const pageMoveUs = elapsedUs(pageStart)

    const jumpStart = process.hrtime.bigint()
    app.jumpToEnd()
    const finalFrame = await app.render()
    const jumpToEndUs = elapsedUs(jumpStart)

    const copyStart = process.hrtime.bigint()
    const copied = app.copySelectedRow()
    const copySelectedRowUs = elapsedUs(copyStart)

    const expectedId = rowId(options.rows - 1)
    const expectedCopy = expectedCopiedRow(options.rows - 1)
    const queryStart = process.hrtime.bigint()
    const snapshot = app.snapshot()
    const frameContainsSelectedRow =
      finalFrame.includes(expectedId) && snapshot.frameContainsSelectedRow
    const semanticOrTestQueryUs = elapsedUs(queryStart)
    const unsafeLeakCount =
      snapshot.lastFrameUnsafeCount +
      unsafeCopyTextCount(copied) +
      unsafeVisibleTextCount(firstFrame) +
      0

    return {
      mountUs,
      firstRenderUs,
      arrowMoveUs,
      pageMoveUs,
      jumpToEndUs,
      copySelectedRowUs,
      semanticOrTestQueryUs,
      rssDeltaBytes: Math.max(0, rssAfterMount - rssBefore),
      rowCount: snapshot.rowCount,
      visibleWindowRows: snapshot.visibleWindowRows,
      visibleStart: snapshot.visibleStart,
      visibleEnd: snapshot.visibleEnd,
      selectedRow: snapshot.selectedRow,
      selectedRowId: snapshot.selectedRowId,
      visibleWindowBounded:
        snapshot.visibleWindowRows <=
        visibleDataCapacity(options.terminalRows),
      selectionCorrect:
        snapshot.selectedRow === options.rows - 1 &&
        snapshot.selectedRowId === expectedId &&
        frameContainsSelectedRow,
      copyExact: copied === expectedCopy && unsafeCopyTextCount(copied) === 0,
      unsafeLeakCount,
      firstFrameUnsafeCount: unsafeVisibleTextCount(firstFrame),
    }
  } finally {
    app?.destroy()
  }
}

function buildArtifact(options, samples, runRssDeltaBytes) {
  const capturedAt = new Date()
  const runId = `opentui-sb3-datatable-${capturedAt
    .toISOString()
    .replace(/\.\d{3}Z$/, 'Z')
    .replaceAll(':', '-')}`
  const last = samples[samples.length - 1]
  const allVisibleBounded = samples.every((sample) => sample.visibleWindowBounded)
  const allSelectionCorrect = samples.every((sample) => sample.selectionCorrect)
  const allCopyExact = samples.every((sample) => sample.copyExact)
  const appLines = sourceLineCount(new URL('../lib/table_app.js', import.meta.url))
  const benchmarkLines = sourceLineCount(import.meta.url)
  const testLines = sourceLineCount(
    new URL('../test/table_app.test.js', import.meta.url),
  )

  return {
    schemaVersion: SCHEMA_VERSION,
    kind: 'fleuryPeerBenchmarkRun',
    runId,
    peerId: PEER_ID,
    scenarioId: SCENARIO_ID,
    capturedAt: capturedAt.toISOString(),
    source: {
      name: PEER_NAME,
      version: PEER_VERSION,
      url: PEER_URL,
    },
    environment: {
      machine: commandOutput('hostname') ?? 'unknown',
      operatingSystem: process.platform,
      operatingSystemVersion:
        commandOutput('uname', ['-a']) ?? process.platform,
      runtime: `${runtimeVersion()} / OpenTUI ${PEER_VERSION}`,
      terminalMode: 'opentui-test-renderer-memory',
      terminalSize: {
        columns: options.terminalColumns,
        rows: options.terminalRows,
      },
    },
    fixture: {
      workingDirectory: 'peer-fixtures/opentui/sb3_datatable',
      command: [
        'npm',
        'run',
        'benchmark',
        '--',
        `--warmup=${options.warmupIterations}`,
        `--iterations=${options.measuredIterations}`,
        `--rows=${options.rows}`,
        '--json',
      ],
      warmupIterations: options.warmupIterations,
      measuredIterations: options.measuredIterations,
    },
    metrics: {
      mountUs: stats(samples.map((sample) => sample.mountUs)),
      firstRenderUs: stats(samples.map((sample) => sample.firstRenderUs)),
      arrowMoveUs: stats(samples.map((sample) => sample.arrowMoveUs)),
      pageMoveUs: stats(samples.map((sample) => sample.pageMoveUs)),
      jumpToEndUs: stats(samples.map((sample) => sample.jumpToEndUs)),
      copySelectedRowUs: stats(samples.map((sample) => sample.copySelectedRowUs)),
      semanticOrTestQueryUs: stats(
        samples.map((sample) => sample.semanticOrTestQueryUs),
      ),
      rssDeltaBytes: Math.max(
        runRssDeltaBytes,
        ...samples.map((sample) => sample.rssDeltaBytes),
      ),
      lineOfCodeCount: appLines,
      benchmarkLineOfCodeCount: benchmarkLines,
      testLineOfCodeCount: testLines,
      rowCount: options.rows,
      observedRowCount: last.rowCount,
      visibleWindowRowEstimate: last.visibleWindowRows,
      visibleRangeStart: last.visibleStart,
      visibleRangeEnd: last.visibleEnd,
      finalSelectedRow: last.selectedRow,
      finalSelectedRowId: last.selectedRowId,
      unsafeLeakCount: Math.max(...samples.map((sample) => sample.unsafeLeakCount)),
    },
    correctness: [
      {
        gate: 'visible window stays bounded',
        pass: allVisibleBounded,
        evidence:
          'Fixture-owned visible-row slice stayed within the OpenTUI test-renderer height.',
      },
      {
        gate: 'selection is correct after jump',
        pass: allSelectionCorrect,
        evidence: `Jump-to-end selected ${rowId(
          options.rows - 1,
        )} and rendered it in the OpenTUI character frame.`,
      },
      {
        gate: 'copy/export is sanitized and exact',
        pass: allCopyExact,
        evidence:
          'Selected row TSV matched generated source row and contained no escape/control/secret leakage.',
      },
    ],
    ergonomics: {
      lineOfCodeCount: appLines,
      benchmarkLineOfCodeCount: benchmarkLines,
      testLineOfCodeCount: testLines,
      appFile: 'lib/table_app.js',
      benchmarkFile: 'bin/sb3_datatable_benchmark.js',
      testFile: 'test/table_app.test.js',
      peerOwnedTextTableRenderable: true,
      appOwnedRetainedRows: true,
      appOwnedVisibleRowSlicing: true,
      appOwnedSelectionState: true,
      appOwnedCopyExport: true,
      semanticGraphAvailable: false,
      testQueryViaFrameAndAppState: true,
    },
    artifacts: [
      {
        kind: 'source',
        path: 'peer-fixtures/opentui/sb3_datatable/lib/table_app.js',
      },
      {
        kind: 'benchmark',
        path: 'peer-fixtures/opentui/sb3_datatable/bin/sb3_datatable_benchmark.js',
      },
      {
        kind: 'test',
        path: 'peer-fixtures/opentui/sb3_datatable/test/table_app.test.js',
      },
    ],
    notes: [
      'This is an OpenTUI test-renderer memory fixture, not a real-terminal run.',
      'OpenTUI 0.3.1 supplies TextTableRenderable, styled text chunks, and the native-backed test renderer.',
      'The fixture owns retained row generation, visible-row slicing, navigation policy, selected-row copy/export, and state query because this scenario needs app behavior around the OpenTUI table renderer.',
      'Timing is local-machine evidence and should not be used as a public superiority claim without repeated peer runs.',
    ],
  }
}

function parseArgs(args) {
  const options = {
    warmupIterations: DEFAULT_WARMUPS,
    measuredIterations: DEFAULT_ITERATIONS,
    rows: DEFAULT_ROWS,
    terminalColumns: DEFAULT_COLUMNS,
    terminalRows: DEFAULT_TERMINAL_ROWS,
    printJson: false,
    outputPath: null,
  }

  for (const arg of args) {
    if (arg === '--json') {
      options.printJson = true
    } else if (arg.startsWith('--warmup=')) {
      options.warmupIterations = parseNonNegativeInteger(arg, 'warmup')
    } else if (arg.startsWith('--iterations=')) {
      options.measuredIterations = parsePositiveInteger(arg, 'iterations')
    } else if (arg.startsWith('--rows=')) {
      options.rows = parsePositiveInteger(arg, 'rows')
    } else if (arg.startsWith('--size=')) {
      const value = arg.slice('--size='.length)
      const [columns, rows] = value.split('x')
      options.terminalColumns = parsePositiveInteger(
        `--columns=${columns}`,
        'columns',
      )
      options.terminalRows = parsePositiveInteger(`--rows=${rows}`, 'rows')
    } else if (arg.startsWith('--output=')) {
      options.outputPath = arg.slice('--output='.length)
    } else {
      throw new Error(`unknown argument: ${arg}`)
    }
  }

  if (options.measuredIterations <= 0) {
    throw new Error('--iterations must be positive')
  }
  if (options.rows <= 0) {
    throw new Error('--rows must be positive')
  }
  return options
}

function parsePositiveInteger(arg, label) {
  const value = Number.parseInt(arg.slice(arg.indexOf('=') + 1), 10)
  if (!Number.isFinite(value) || value <= 0) {
    throw new Error(`--${label} must be a positive integer`)
  }
  return value
}

function parseNonNegativeInteger(arg, label) {
  const value = Number.parseInt(arg.slice(arg.indexOf('=') + 1), 10)
  if (!Number.isFinite(value) || value < 0) {
    throw new Error(`--${label} must be a non-negative integer`)
  }
  return value
}

function elapsedUs(start) {
  return Number((process.hrtime.bigint() - start) / 1000n)
}

function stats(values) {
  const ordered = [...values].sort((a, b) => a - b)
  if (ordered.length === 0) {
    return { min: 0, median: 0, p95: 0, p99: 0, max: 0, samples: 0 }
  }
  return {
    min: ordered[0],
    median: percentile(ordered, 0.5),
    p95: percentile(ordered, 0.95),
    p99: percentile(ordered, 0.99),
    max: ordered[ordered.length - 1],
    samples: ordered.length,
  }
}

function percentile(ordered, fraction) {
  if (ordered.length === 1) return ordered[0]
  const index = Math.ceil((ordered.length - 1) * fraction)
  return ordered[Math.min(index, ordered.length - 1)]
}

function sourceLineCount(url) {
  const path = fileURLToPath(url)
  const text = readFileSync(path, 'utf8')
  return text
    .split(/\r?\n/)
    .map((line) => line.trim())
    .filter((line) => line.length > 0 && !line.startsWith('//')).length
}

function commandOutput(command, args = []) {
  try {
    return execFileSync(command, args, { encoding: 'utf8' }).trim()
  } catch {
    return null
  }
}

function runtimeVersion() {
  const bunVersion = globalThis.Bun?.version
  if (bunVersion) return `Bun ${bunVersion}`
  return `Node ${process.version}`
}

File: peer-fixtures/nocterm/sb3_datatable/bin/sb3_datatable_benchmark.dart
import 'dart:convert';
import 'dart:io';
import 'dart:math' as math;

import 'package:fleury_peer_nocterm_sb3_datatable/table_app.dart';
import 'package:nocterm/nocterm.dart';

const _schemaVersion = 1;
const _peerId = 'nocterm';
const _peerName = 'Nocterm';
const _peerVersion = '0.6.0';
const _peerUrl = 'https://pub.dev/packages/nocterm';
const _scenarioId = 'SB.3';
const _defaultWarmups = 2;
const _defaultIterations = 5;
const _defaultRows = 100000;
const _defaultSize = Size(120, 32);

Future<void> main(List<String> args) async {
  final options = _Options.parse(args);

  for (var i = 0; i < options.warmupIterations; i += 1) {
    await _runSample(options);
  }

  final samples = <_Sample>[];
  for (var i = 0; i < options.measuredIterations; i += 1) {
    samples.add(await _runSample(options));
  }

  final artifact = _buildArtifact(options, samples);
  final jsonText = const JsonEncoder.withIndent('  ').convert(artifact);
  if (options.outputPath != null) {
    final output = File(options.outputPath!);
    output.parent.createSync(recursive: true);
    output.writeAsStringSync('$jsonText\n');
  }

  if (options.printJson) {
    stdout.writeln(jsonText);
    return;
  }

  final metrics = artifact['metrics']! as Map<String, Object?>;
  stdout.writeln('Nocterm SB.3 data table fixture');
  stdout.writeln('Run: ${artifact['runId']}');
  stdout.writeln('Rows: ${options.rows}');
  stdout.writeln('Iterations: ${options.measuredIterations}');
  stdout.writeln('pageMoveUs p95: ${(metrics['pageMoveUs']! as Map)['p95']}');
  stdout.writeln(
    'copySelectedRowUs p95: '
    '${(metrics['copySelectedRowUs']! as Map)['p95']}',
  );
  if (options.outputPath != null) {
    stdout.writeln('Saved ${options.outputPath}');
  }
}

Future<_Sample> _runSample(_Options options) async {
  final rssBefore = ProcessInfo.currentRss;
  final tester = await NoctermTester.create(size: options.size);
  try {
    final mount = Stopwatch()..start();
    await tester.pumpComponent(
      Sb3NoctermDataTable(
        rowCount: options.rows,
        width: options.size.width.toInt(),
        height: options.size.height.toInt(),
      ),
    );
    mount.stop();
    final state = tester.findState<Sb3NoctermDataTableState>();

    final firstRender = Stopwatch()..start();
    final firstView = tester.terminalState.getText();
    firstRender.stop();

    final arrow = Stopwatch()..start();
    state.arrowDown();
    await tester.pump();
    arrow.stop();
    final arrowState = state.snapshot();
    final arrowView = tester.terminalState.getText();

    final page = Stopwatch()..start();
    state.pageDown();
    await tester.pump();
    page.stop();
    final pageState = state.snapshot();
    final pageView = tester.terminalState.getText();

    final jump = Stopwatch()..start();
    state.jumpToEnd();
    await tester.pump();
    jump.stop();
    final finalState = state.snapshot();
    final finalView = tester.terminalState.getText();

    final copy = Stopwatch()..start();
    state.copySelectedRow();
    copy.stop();
    final copied = state.lastCopiedText;

    final expectedFinalIndex = options.rows - 1;
    final expectedFinalId = rowId(expectedFinalIndex);
    final query = Stopwatch()..start();
    final selectedVisible = tester.terminalState.containsText(expectedFinalId);
    final unsafeLeakCount =
        finalState.unsafeArtifactLeakCount +
        unsafeVisibleTextCount(firstView) +
        unsafeVisibleTextCount(arrowView) +
        unsafeVisibleTextCount(pageView) +
        unsafeVisibleTextCount(finalView);
    query.stop();

    return _Sample(
      mountUs: mount.elapsedMicroseconds,
      firstRenderUs: firstRender.elapsedMicroseconds,
      arrowMoveUs: arrow.elapsedMicroseconds,
      pageMoveUs: page.elapsedMicroseconds,
      jumpToEndUs: jump.elapsedMicroseconds,
      copySelectedRowUs: copy.elapsedMicroseconds,
      semanticOrTestQueryUs: query.elapsedMicroseconds,
      rssDeltaBytes: math.max(0, ProcessInfo.currentRss - rssBefore),
      rowCount: finalState.rowCount,
      visibleWindowRows: finalState.visibleWindowRows,
      visibleStart: finalState.visibleStart,
      visibleEnd: finalState.visibleEnd,
      selectedRow: finalState.selectedRow,
      selectedRowId: finalState.selectedRowId,
      unsafeArtifactLeakCount: unsafeLeakCount,
      visibleWindowBounded:
          finalState.visibleWindowRows <= state.visibleCapacity(),
      selectionCorrect:
          finalState.selectedRow == expectedFinalIndex &&
          finalState.selectedRowId == expectedFinalId &&
          selectedVisible,
      copyExact:
          copied == expectedSelectedTsv(expectedFinalIndex) &&
          unsafeCopyTextCount(copied) == 0,
      arrowCorrect: arrowState.selectedRow == 1 && arrowView.contains(rowId(1)),
      pageCorrect:
          pageState.selectedRow > 1 &&
          pageView.contains(pageState.selectedRowId),
    );
  } finally {
    tester.dispose();
  }
}

Map<String, Object?> _buildArtifact(_Options options, List<_Sample> samples) {
  final packageRoot = _packageRoot();
  final appLines = _sourceLineCount('${packageRoot.path}/lib/table_app.dart');
  final benchmarkLines = _sourceLineCount(
    '${packageRoot.path}/bin/sb3_datatable_benchmark.dart',
  );
  final testLines = _sourceLineCount(
    '${packageRoot.path}/test/table_benchmark_test.dart',
  );
  final capturedAt = DateTime.now().toUtc();
  final runId = 'nocterm-sb3-datatable-${_timestampForId(capturedAt)}';
  final unsafeLeakCount = samples
      .map((sample) => sample.unsafeArtifactLeakCount)
      .reduce(math.max);
  final last = samples.last;

  return <String, Object?>{
    'schemaVersion': _schemaVersion,
    'kind': 'fleuryPeerBenchmarkRun',
    'runId': runId,
    'peerId': _peerId,
    'scenarioId': _scenarioId,
    'capturedAt': capturedAt.toIso8601String(),
    'source': <String, Object?>{
      'name': _peerName,
      'version': _peerVersion,
      'url': _peerUrl,
    },
    'environment': <String, Object?>{
      'machine': Platform.localHostname,
      'operatingSystem': Platform.operatingSystem,
      'operatingSystemVersion': Platform.operatingSystemVersion,
      'runtime': Platform.version,
      'terminalMode': 'nocterm-test-harness',
      'terminalSize': <String, Object?>{
        'columns': options.size.width.toInt(),
        'rows': options.size.height.toInt(),
      },
    },
    'fixture': <String, Object?>{
      'workingDirectory': 'peer-fixtures/nocterm/sb3_datatable',
      'command': <String>[
        'dart',
        'run',
        'bin/sb3_datatable_benchmark.dart',
        '--warmup=${options.warmupIterations}',
        '--iterations=${options.measuredIterations}',
        '--rows=${options.rows}',
        '--json',
      ],
      'warmupIterations': options.warmupIterations,
      'measuredIterations': options.measuredIterations,
    },
    'metrics': <String, Object?>{
      'mountUs': _stats(samples.map((sample) => sample.mountUs)),
      'firstRenderUs': _stats(samples.map((sample) => sample.firstRenderUs)),
      'arrowMoveUs': _stats(samples.map((sample) => sample.arrowMoveUs)),
      'pageMoveUs': _stats(samples.map((sample) => sample.pageMoveUs)),
      'jumpToEndUs': _stats(samples.map((sample) => sample.jumpToEndUs)),
      'copySelectedRowUs': _stats(
        samples.map((sample) => sample.copySelectedRowUs),
      ),
      'semanticOrTestQueryUs': _stats(
        samples.map((sample) => sample.semanticOrTestQueryUs),
      ),
      'rssDeltaBytes': samples
          .map((sample) => sample.rssDeltaBytes)
          .reduce(math.max),
      'lineOfCodeCount': appLines,
      'benchmarkLineOfCodeCount': benchmarkLines,
      'testLineOfCodeCount': testLines,
      'rowCount': options.rows,
      'observedRowCount': last.rowCount,
      'visibleWindowRowEstimate': last.visibleWindowRows,
      'visibleRangeStart': last.visibleStart,
      'visibleRangeEnd': last.visibleEnd,
      'finalSelectedRow': last.selectedRow,
      'finalSelectedRowId': last.selectedRowId,
      'unsafeArtifactLeakCount': unsafeLeakCount,
    },
    'correctness': <Object?>[
      <String, Object?>{
        'gate': 'visible window stays bounded',
        'pass': samples.every((sample) => sample.visibleWindowBounded),
        'evidence':
            'Nocterm ListView rendered a bounded viewport around fixture-owned table rows.',
      },
      <String, Object?>{
        'gate': 'selection is correct after jump',
        'pass': samples.every((sample) => sample.selectionCorrect),
        'evidence':
            'Jump-to-end selected ${rowId(options.rows - 1)} and rendered it in terminal state.',
      },
      <String, Object?>{
        'gate': 'copy/export is sanitized and exact',
        'pass':
            samples.every((sample) => sample.copyExact) && unsafeLeakCount == 0,
        'evidence':
            'Selected row TSV matched generated source row and contained no escape/control/secret leakage.',
      },
    ],
    'ergonomics': <String, Object?>{
      'lineOfCodeCount': appLines,
      'benchmarkLineOfCodeCount': benchmarkLines,
      'testLineOfCodeCount': testLines,
      'appFile': 'lib/table_app.dart',
      'benchmarkFile': 'bin/sb3_datatable_benchmark.dart',
      'testFile': 'test/table_benchmark_test.dart',
      'peerOwnedListView': true,
      'peerOwnedScrollController': true,
      'peerOwnedText': true,
      'appOwnedTableFormatting': true,
      'appOwnedRetainedRows': true,
      'appOwnedSelectionState': true,
      'appOwnedCopyExport': true,
      'semanticGraphAvailable': false,
      'testQueryViaTerminalStateAndAppState': true,
    },
    'artifacts': <Object?>[
      <String, Object?>{
        'kind': 'source',
        'path': 'peer-fixtures/nocterm/sb3_datatable/lib/table_app.dart',
      },
      <String, Object?>{
        'kind': 'benchmark',
        'path':
            'peer-fixtures/nocterm/sb3_datatable/bin/sb3_datatable_benchmark.dart',
      },
      <String, Object?>{
        'kind': 'test',
        'path':
            'peer-fixtures/nocterm/sb3_datatable/test/table_benchmark_test.dart',
      },
    ],
    'notes': <String>[
      'This is a Nocterm test-harness peer fixture, not a real-terminal run.',
      'Nocterm 0.6.0 supplies ListView.builder, lazy list rendering, ScrollController, Text rendering, and terminal-state test queries.',
      'Nocterm 0.6.0 does not expose a dedicated DataTable widget in this package source, so table formatting, retained rows, selection, visible-window policy, and copy/export are app-owned fixture code.',
      'Nocterm exposes terminal/app state for this fixture, not a Fleury-style semantic app graph.',
      'Timing is local-machine evidence and should not be used as a public superiority claim without repeated peer runs.',
    ],
  };
}

Map<String, Object?> _stats(Iterable<int> values) {
  final sorted = values.toList()..sort();
  if (sorted.isEmpty) {
    throw StateError('Cannot summarize empty metric samples.');
  }
  return <String, Object?>{
    'min': sorted.first,
    'median': _percentile(sorted, 0.50),
    'p95': _percentile(sorted, 0.95),
    'p99': _percentile(sorted, 0.99),
    'max': sorted.last,
    'samples': sorted.length,
  };
}

int _percentile(List<int> sorted, double percentile) {
  final index = ((sorted.length * percentile).ceil() - 1).clamp(
    0,
    sorted.length - 1,
  );
  return sorted[index];
}

Directory _packageRoot() {
  final script = File(Platform.script.toFilePath()).absolute;
  return script.parent.parent;
}

int _sourceLineCount(String path) {
  final lines = File(path).readAsLinesSync();
  var count = 0;
  for (final line in lines) {
    final trimmed = line.trim();
    if (trimmed.isEmpty || trimmed.startsWith('//')) continue;
    count += 1;
  }
  return count;
}

String _timestampForId(DateTime value) {
  String two(int value) => value.toString().padLeft(2, '0');
  return '${value.year}-${two(value.month)}-${two(value.day)}T'
      '${two(value.hour)}-${two(value.minute)}-${two(value.second)}Z';
}

final class _Sample {
  const _Sample({
    required this.mountUs,
    required this.firstRenderUs,
    required this.arrowMoveUs,
    required this.pageMoveUs,
    required this.jumpToEndUs,
    required this.copySelectedRowUs,
    required this.semanticOrTestQueryUs,
    required this.rssDeltaBytes,
    required this.rowCount,
    required this.visibleWindowRows,
    required this.visibleStart,
    required this.visibleEnd,
    required this.selectedRow,
    required this.selectedRowId,
    required this.unsafeArtifactLeakCount,
    required this.visibleWindowBounded,
    required this.selectionCorrect,
    required this.copyExact,
    required this.arrowCorrect,
    required this.pageCorrect,
  });

  final int mountUs;
  final int firstRenderUs;
  final int arrowMoveUs;
  final int pageMoveUs;
  final int jumpToEndUs;
  final int copySelectedRowUs;
  final int semanticOrTestQueryUs;
  final int rssDeltaBytes;
  final int rowCount;
  final int visibleWindowRows;
  final int visibleStart;
  final int visibleEnd;
  final int selectedRow;
  final String selectedRowId;
  final int unsafeArtifactLeakCount;
  final bool visibleWindowBounded;
  final bool selectionCorrect;
  final bool copyExact;
  final bool arrowCorrect;
  final bool pageCorrect;
}

final class _Options {
  const _Options({
    required this.warmupIterations,
    required this.measuredIterations,
    required this.rows,
    required this.size,
    required this.printJson,
    required this.outputPath,
  });

  final int warmupIterations;
  final int measuredIterations;
  final int rows;
  final Size size;
  final bool printJson;
  final String? outputPath;

  static _Options parse(List<String> args) {
    var warmupIterations = _defaultWarmups;
    var measuredIterations = _defaultIterations;
    var rows = _defaultRows;
    var size = _defaultSize;
    var printJson = false;
    String? outputPath;

    for (final arg in args) {
      if (arg == '--json') {
        printJson = true;
      } else if (arg.startsWith('--warmup=')) {
        warmupIterations = _parseNonNegativeInt(arg, '--warmup=');
      } else if (arg.startsWith('--iterations=')) {
        measuredIterations = _parsePositiveInt(arg, '--iterations=');
      } else if (arg.startsWith('--rows=')) {
        rows = _parsePositiveInt(arg, '--rows=');
      } else if (arg.startsWith('--size=')) {
        size = _parseSize(arg.substring('--size='.length));
      } else if (arg.startsWith('--output=')) {
        outputPath = arg.substring('--output='.length);
      } else {
        throw ArgumentError('Unknown argument: $arg');
      }
    }

    return _Options(
      warmupIterations: warmupIterations,
      measuredIterations: measuredIterations,
      rows: rows,
      size: size,
      printJson: printJson,
      outputPath: outputPath,
    );
  }

  static int _parsePositiveInt(String arg, String prefix) {
    final value = int.tryParse(arg.substring(prefix.length));
    if (value == null || value <= 0) {
      throw ArgumentError('$prefix requires a positive integer.');
    }
    return value;
  }

  static int _parseNonNegativeInt(String arg, String prefix) {
    final value = int.tryParse(arg.substring(prefix.length));
    if (value == null || value < 0) {
      throw ArgumentError('$prefix requires a non-negative integer.');
    }
    return value;
  }

  static Size _parseSize(String value) {
    final parts = value.split('x');
    if (parts.length != 2) {
      throw ArgumentError('--size requires COLUMNSxROWS.');
    }
    final columns = int.tryParse(parts[0]);
    final rows = int.tryParse(parts[1]);
    if (columns == null || rows == null || columns <= 0 || rows <= 0) {
      throw ArgumentError('--size requires positive integer dimensions.');
    }
    return Size(columns.toDouble(), rows.toDouble());
  }
}
